lib/SILOptimizer/Utils/CanonicalizeInstruction.cpp

```diff
@@ -64,16 +95,121 @@ simplifyAndReplace(SILInstruction *inst, CanonicalizeInstruction &pass) {
       [&pass](SILInstruction *deleted) { pass.killInstruction(deleted); });
 
   // Push the new instruction and any users onto the worklist.
-  pass.newUsers(result);
+  pass.notifyHasNewUsers(result);
   return nextII;
 }
+
+//===----------------------------------------------------------------------===//
+//                        Canonicalize Memory Operations
+//===----------------------------------------------------------------------===//
+
+// Given a load with multiple struct_extracts/tuple_extracts and no other uses,
+// canonicalize the load into several (struct_element_addr (load)) pairs.
+//
+// (struct_extract (load %base))
+//   ->
+// (load (struct_element_addr %base, #field)
+static SILBasicBlock::iterator
+splitAggregateLoad(LoadInst *loadInst, CanonicalizeInstruction &pass) {
+  // Keep track of the next iterator after any newly added or to-be-deleted
+  // instructions. This must be valid regardless of whether the pass immediately
+  // deletes the instructions or simply records them for later deletion.
+  auto nextII = std::next(loadInst->getIterator());
+
+  switch (loadInst->getOwnershipQualifier()) {
+  case LoadOwnershipQualifier::Unqualified:
+  case LoadOwnershipQualifier::Trivial:
+    break;
+  case LoadOwnershipQualifier::Take:
+  case LoadOwnershipQualifier::Copy:
+    // Skip ahead if the ownership qualifier isn't handled.
+    //
+    // FIXME: To handle Copy/Take, the code below needs to properly peek through
+    // and destroy borrow scopes.
+    return nextII;
+  }
+  struct ProjInstPair {
+    Projection proj;
+    SingleValueInstruction *svi;
+
+    // When sorting, just look at the projection and ignore the instruction.
```

Comment:
I know that this is just moving some stuff around, but can you add to the comment here saying that not doing this results in non-determinism since it is possible to have multiple projections that are the same with different instruction pointers. So when we sort, the order may change subtly between compiler invocations == unhappiness. I just want to make sure that this is memorialized in the comment to prevent someone from removing this due to lack of context.

Reply:
Yes, done.